### ISSUE TITLE
Any way to reduce container size?

### ISSUE THREAD

**Issue by nsheff** [pepkit/pephub#109]
Up until pephub 0.4.0, the container size was under 500 mb. With 0.5.0 it exploded to > 5 gb. This was due to the need to include, I think, all of tensorflow for sentence transformers.

We should consider if there's a way we can reduce these installations to keep the container to a more reasonable size, since I believe this is leading to some deployment challenges.

**Comment by nleroy917** [comment]
huggingface has [docker containers for everything](https://hub.docker.com/r/huggingface/transformers-pytorch-gpu), but they are just as large (6-8GB).

If you read through [this issue thread](https://github.com/pytorch/pytorch/issues/17621), it seems like there are ways to get the size down if you compile for a specific architecture.

> You can get a smaller binary by building from source and only specifying the architecture that you need for your GPU, but for the public binaries we just want one to fit them all.

Even smaller if you compile the CPU-inference code. This works for us since we don't take advantage of the GPU capabilities since we are just using pre-trained models. Even more, there is this comment:

> Just pointing out that there are public CPU builds that are of a more reasonable size (~180 MB), unfortunately not on pypi but pytorch's own repo. 
> e.g. you could install it using the following requirements.txt:
```
-f https://download.pytorch.org/whl/torch_stable.html
torch==1.9.1+cpu
torchvision==0.10.1+cpu
```

**Comment by HeroadZ** [comment]
tensorflow/serving only takes ~500MB while torchserve takes about 4GB. This is really a problom for people who want to deploy pytorch model to production.

**Comment by nleroy917** [comment]
Interestingly, the operating system on which you build the container seems to determine the size of the container.

**On macOS:** ~1.5GB
**On local Ubuntu (me):** ~12.5GB
**On remote Ubuntu (GitHub Runner):** ~12.5GB
**On local Ubuntu (Alex):** ~8.5GB

I dont know *why* this would be the case, however. Another thing that is a clue, is it's always the `pip install` layer that drives up the size. Namely, `pytorch`.

**On macOS:**
```
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
36cccc0201e5   13 minutes ago   CMD ["uvicorn" "pephub.main:app" "--host" "0…   0B        buildkit.dockerfile.v0
<missing>      13 minutes ago   RUN /bin/sh -c pip install -r requirements/r…   1.05GB    buildkit.dockerfile.v0
```

**On local Ubuntu (me):**
```
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
636520bf5cf1   9 minutes ago    /bin/sh -c #(nop)  CMD ["uvicorn" "pephub.ma…   0B        
95046ecb7c4e   9 minutes ago    /bin/sh -c pip install -r requirements/requi…   11.8GB 
```

**On remote Ubuntu (GitHub runner):**
```
IMAGE          CREATED             CREATED BY                                      SIZE      COMMENT
e44f8d15cd41   About an hour ago   /bin/sh -c #(nop)  CMD ["uvicorn" "pephub.ma…   0B        
<missing>      About an hour ago   /bin/sh -c pip install -r requirements/requi…   11.8GB 
```

**Comment by nleroy917** [comment]
I think I figured it out: when specifying `+cpu` torch dependencies you **must** pin the version. I was not doing this.
**Before:**
```
# requirements-all.txt
-f https://download.pytorch.org/whl/torch_stable.html
torch>=1.13.0+cpu
torchvision>=0.14.0+cpu
```
**After:**
```
# requirements-all.txt
-f https://download.pytorch.org/whl/torch_stable.html
torch==1.13.1+cpu
torchvision==0.14.1+cpu
```

Note the `==` v `>=`. Making this switch drops the image size on Ubuntu drastically (~12.5 GB --> ~1.5GB). Interestingly, there doesn't seem to exist a `torch==1.13.1+cpu` for macOS natively. If you try to install it, it fails. Maybe this is a clue as to why the image size was exploding with Ubuntu originally and not macOS.

Additionally, I am using the `--no-cache-dir` flag on install as recommended by that [earlier thread](https://github.com/pytorch/pytorch/issues/17621).

**Comment by HeroadZ** [comment]
it seems like that torch2.0 is smaller than torch1.x

**Comment by nleroy917** [comment]
This is now solved in the latest release (`0.7.x`)
